documentation/adr/0017-separate-absolute-and-relative-indexing-APIs-in-pixel-array.md

```diff
@@ -0,0 +1,70 @@
+[<-previous](./0016-use-double-array-in-memory-for-pixel-data.md) |
+next->
+
+# 17 - Separate absolute and relative indexing APIs in pixel array
+
+Date: 2020-Oct-15
+
+## Status
+
+Accepted
+
+## Context
+
+The pixel array within an SQW object can be too large to fit into memory.
+To avoid running out of memory, the object holding the pixel array can be
+file-backed.
+This means that only a "page" of the pixel array is loaded into memory at any
+one time.
+
+Therefore two possible ways to index into the pixel array exist:
+
+1. **Absolute index**:
+_The position of the pixel in the full, file-backed, pixel array_.
+
+2. **Relative index**:
+_The position of the pixel in the currently loaded page of pixel data._
+
+## Decision
+
+There will be two separate APIs for accessing data,
+these APIs distinguish between the two types of indexing.
+
+1. **Perform absolute indexing  using `get_` methods:**
+
+    Obtaining a subset of pixels or pixel data by absolute index will be
+    possible using a `get_` method.
+    For example, the following will retrieve pixels 100-200 by absolute index:
+
+    ```matlab
+    pixels.get_pixels(100:200)
+    ```
+
+    Similarly, to retrieve a range of data from particular pixel array fields:
+
+    ```matlab
+    pixels.get_data(100:200, {'signal', 'variance'})
+    ```
+
+2. **Perform relative indexing using attribute:**
+
+    Obtaining pixel data using an attribute will return just the data for the
+    currently cached page.
+    Hence, indexing into these attributes will be relative.
+    For example, the following will retrieve the signal values of pixels 10-20
+    in the currently cached page:
+
+    ```matlab
+    pixels.signal(10:20)
+    ```
```

Comment:
There is nothing here on the need for put_ methods - writing data into the structures and onto permanent disk. At least worth stating that this will need to be addressed.

Reply:
There are no public `put_` methods on the `PixelData` objects that write to files. I don't think there's a requirement to be able to write pixel data to file without the rest of the accompanying SQW object. It's the responsibility of the IO functions to write the `PixelData` object's underlying data into the SQW file.

As for `put_` (or `set_`?) methods that sets data within objects, I think you're probably right, we should mention those. At the moment there are no `set_` equivalents of `get_pixels` or `get_data` - we haven't needed them. Data can be set using the attributes (relative indexing), e.g. `pix.signal(1:10) = 0`.